helpers/Mixins.cfm

```diff
@@ -31,7 +31,40 @@
 	 *
 	 * @return HTML of the hidden field (csrf)
 	 */
-	function csrf( string key='', boolean forceNew=false ){
+	string function csrf( string key='', boolean forceNew=false ){
+		return "<input type='hidden' name='csrf' id='csrf' value='#csrfToken( argumentCollection=arguments )#'>";
+	}
+
+	/**
+	 * Generate a random token in a hidden form element and javascript that will refresh the page automatically when the token expires
+	 * 
+	 * @key A random token is generated for the key provided. CFID is the default
+	 * @forceNew If set to true, a new token is generated every time the function is called. If false, in case a token exists for the key, the same key is returned.
+	 *
+	 * @return HTML of the hidden field (csrf)
+	 */
+	string function csrfField( string key='', boolean forceNew=false ){
+		if (!event.getPrivateValue('csrfFieldGen', false)){
+			savecontent variable="block" {
+				writeOutput("
+				<script>
+					var check = Date.now();
+					var timeout = #getModuleSettings( 'cbcsrf' ).rotationTimeout#; // expiration is set in /config/coldbox.cfc moduleSettings.cbcsrf.rotationTimeout
+					document.addEventListener('focus',function(){
```

Comment:
This event listener will only fire once though, once the document has focus, but it won't be checking for the timeout every time.  Shouldn't this be a setTimeout() where every x amount of seconds it checks?

I am not sure if this is something that should be core because it assumes too much. What If I just wanted to regenerate a key via ajax instead of reloading the entire form? What happens to the form elements already written?

Reply:
Any thoughts @bdw429s

Reply:
setTimeout() using the rotation timeout setting would work, but the downside is that it will be refreshing the key every x minutes even in cases where the user has 100 tabs open and the form is sitting idle for days.  In most cases the user will be submitting the form immediately.  I get burned when I let a login sit after switching tabs, then switch back at some point after the rotation timeout has expired and try to log in.  In those cases focus works perfectly as it fires whether you are switching tabs or switching back to the browser with your page already loaded, but does nothing until that page is focused.  It's easier to just reload the page to get a fresh token in these cases than it is to intercept the invalid token exception and try to explain to the user what went wrong and/or what they need to do next.

You wouldn't use this in situations where you're using ajax to load and submit the form - you would use the endpoint instead.  And if you want to store a token in a field rather than grabbing a fresh one from the endpoint immediately before submitting, (which is what I thought the intent was, as you could determine whether the user is logged in before submitting too), you can use the existing csrf() function to create the hidden field without the refresh.

Reply:
Of course, I could do a setTimeout() that checks to see if the page is focused before reloading too...